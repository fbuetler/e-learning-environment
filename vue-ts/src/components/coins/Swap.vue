<template>
  <div>
    <slot name="animation" :animationSteps="animationSteps" />
    <div>Stelle die gleiche Summe mit weniger Münzen dar</div>
    <div id="coins" class="flex-item flex-center flex-col flex-flex card">
      <slot>
        <div class="flex-item flex-center flex-row">
          <div
            v-for="(amount, i) in generatedItems"
            :key="`item-${i}`"
            class="flex-item flex-center flex-col"
          >
            <div
              v-for="j in amount"
              :key="`amount-${j}`"
              :class="items(type)[i].class"
            >
              <img :src="require(`@/assets/${items(type)[i].img}`)" />
            </div>
          </div>
        </div>
      </slot>
    </div>
    <hr />
    <div>Lösung:</div>
    <div
      id="dropzone"
      class="flex-item flex-center flex-col flex-flex dropzone"
      @click="addItem()"
      @dragover.prevent
      @dragend.prevent
      @drop.stop.prevent="addItem()"
    >
      <div v-if="nothingSelected">
        Platziere hier die Münzen
      </div>
      <div v-else>
        <slot>
          <div class="flex-item flex-center flex-row">
            <div
              v-for="(amount, i) in displaySelectedItems"
              :key="`item-${i}`"
              class="flex-item flex-center flex-col"
            >
              <div
                v-for="j in amount"
                :key="`amount-${j}`"
                :class="items(type).reverse()[i].class"
              >
                <img
                  :src="require(`@/assets/${items(type).reverse()[i].img}`)"
                />
              </div>
            </div>
          </div>
        </slot>
      </div>
    </div>
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch"
    >
      <ItemSelection
        :selected="selected"
        :items="items(type).reverse()"
        @selected="selected = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins.vue";
import CoinsMixin, { coinType } from "@/components/coins/CoinsMixin.vue";
import ItemSelection from "@/components/ItemSelection.vue";
import Undo from "@/components/Undo.vue";
import Difficulty from "@/components/Difficulty.vue";

// TODO: (optional) less coins or minimal number of coins? -> both: difficulties

@Component<Swap>({
  components: {
    ItemSelection,
    Undo,
    Difficulty,
  },
})
export default class Swap extends Mixins(GameMixin, CoinsMixin)
  implements GameInterface {
  type = coinType.NORMAL;

  selected: number = null;
  selectedItems: Array<number> = null;
  generatedItems: Array<number> = null;
  animationSteps: Array<string> = null;

  currentDifficultyLevel = 1;
  difficultyLevels = 2;

  isStarted(): boolean {
    return this.generatedItems === null;
  }

  restartGame() {
    this.selected = null;
    this.selectedItems = new Array<number>(this.items(this.type).length).fill(
      0
    );
    do {
      this.generatedItems = this.generateItems(this.type);
    } while (
      this.countCoins(this.generatedItems) ===
      this.countCoins(
        this.calcMinimalAmount(
          this.sumItems(this.type, this.generatedItems),
          this.type
        )
      )
    );
    this.animationSteps = this.getAnimationSteps();
  }

  isCorrect(): boolean {
    return (
      this.sumItems(this.type, this.selectedItems) ===
        this.sumItems(this.type, this.generatedItems) &&
      this.countCoins(this.selectedItems) < this.countCoins(this.generatedItems)
    );
  }

  countCoins(arr: number[]): number {
    return arr.reduce((sum, el) => (sum += el));
  }

  addItem() {
    if (this.selected === null) {
      return;
    }
    Vue.set(
      this.selectedItems,
      this.selected,
      this.selectedItems[this.selected] + 1
    );
    this.selected = null;
  }

  undo() {
    for (let i = 0; i < this.selectedItems.length; i++) {
      Vue.set(this.selectedItems, i, 0);
    }
  }

  getAnimationSteps(): Array<string> {
    const correctSum = this.sumItems(this.type, this.generatedItems);
    const wrongSum = Math.ceil(Math.random() * this.limit);

    return this.mapNumberToActions(wrongSum, this.type)
      .concat(["button-menu-check", "undo"])
      .concat(this.mapNumberToActions(correctSum, this.type))
      .concat(["button-menu-check", "button-menu-next"]);
  }

  get nothingSelected(): boolean {
    return this.selectedItems.every((el) => el === 0);
  }

  get displaySelectedItems(): number[] {
    return this.selectedItems.slice().reverse();
  }
}
</script>

<style scoped>
.dropzone {
  min-height: 5em;
}
</style>
