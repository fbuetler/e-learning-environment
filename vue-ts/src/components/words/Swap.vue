<template>
  <div>
    <div class="word-container">
      <div v-for="element in word" :key="element.id">
        <div
          class="word-char"
          :class="{
            locked: element.locked,
            selected: element.id === selectedChar,
          }"
          @click="swapChar(element.id)"
          draggable
          @dragstart="selectedChar = element.id"
          @dragover.prevent
          @dragend.prevent
          @drop.stop.prevent="swapChar(element.id)"
        >
          {{ element.char }}
        </div>
      </div>
    </div>
    <hr />
    <div class="interaction-container">
      <Undo @undo-operation="undo($event)" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "../Game";
import Undo from "@/components/Undo.vue";
import { LoadWords, wordElement } from "./LoadWords";

@Component<Change>({
  components: {
    Undo,
  },
})
export default class Change extends Mixins(GameMixin) implements GameInterface {
  private dataKey = "change";
  private word: wordElement[] = null;
  private similarWords: string[] = null;
  private selectedChar: number = null;
  private randomIndex: number = null;
  private charSwaped = false;

  isStarted(): boolean {
    return this.word === null;
  }

  restartGame() {
    this.word = LoadWords(this.dataKey)[0];
    this.randomIndex = Math.floor(Math.random() * (this.word.length - 1));
    this.swap(this.word[this.randomIndex], this.word[this.randomIndex + 1]);

    this.selectedChar = null;
    this.charSwaped = false;
  }

  isCorrect(): boolean {
    return this.word.reduce(
      (acc, currVal) => acc && currVal.initialChar === currVal.char,
      true
    );
  }

  swapChar(id: number) {
    if (this.charSwaped) {
      return;
    }
    if (this.selectedChar === null) {
      this.selectedChar = id;
      return;
    }

    this.swap(
      this.word.find((el) => el.id === id),
      this.word.find((el) => el.id === this.selectedChar)
    );

    this.word.forEach((el) => (el.locked = true));
    this.charSwaped = true;
    this.selectedChar = null;
  }

  swap(a: wordElement, b: wordElement) {
    const temp = a.char;
    a.char = b.char;
    b.char = temp;
  }

  undo() {
    this.word.forEach((el) => {
      el.char = el.initialChar;
      el.locked = false;
    });
    this.swap(this.word[this.randomIndex], this.word[this.randomIndex + 1]);
    this.selectedChar = null;
    this.charSwaped = false;
  }
}
</script>

<style scoped></style>
