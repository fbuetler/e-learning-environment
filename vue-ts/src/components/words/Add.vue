<template>
  <div>
    <div class="word-container">
      <div class="word-item" v-for="element in word" :key="element.id">
        <div
          class="placeholder"
          @click="addChar(element.id)"
          @dragover.prevent
          @dragend.prevent
          @drop.stop.prevent="addChar(element.id)"
        >
          ?
        </div>
        <div
          class="word-char"
          :class="{
            locked: element.locked,
            selected: element.id === selectedChar,
          }"
          @click="selectedChar = element.id"
          draggable
          @dragover.prevent
          @dragend.prevent
          @dragstart="dragChar($event, element.id)"
        >
          {{ element.char }}
        </div>
      </div>
      <div
        class="placeholder"
        @click="addChar(word.length)"
        @dragover.prevent
        @dragend.prevent
        @drop.stop.prevent="addChar(word.length)"
      >
        ?
      </div>
    </div>
    <hr />
    <div class="interaction-container">
      <Alphabet
        :selectedChar="selectedChar"
        @char-selected="selectedChar = $event"
      />
      <Trashcan @trashed-element="trashElement($event)" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "../Game";
import Alphabet from "@/components/words/Alphabet.vue";
import Trashcan from "@/components/Trashcan.vue";
import { LoadWords, wordElement } from "./Words";

@Component<Add>({
  components: {
    Alphabet,
    Trashcan,
  },
})
export default class Add extends Mixins(GameMixin) implements GameInterface {
  private dataKey = "add";
  private word: wordElement[] = null;
  private similarWords: string[] = null;
  private selectedChar: string = null;
  private charAdded = false;

  isStarted(): boolean {
    return this.word === null;
  }

  restartGame() {
    [this.word, this.similarWords] = LoadWords(this.dataKey);
    this.word.forEach((el) => (el.locked = true));
  }

  isCorrect() {
    return this.similarWords.includes(this.word.map((el) => el.char).join(""));
  }

  addChar(addBefore: number) {
    if (this.selectedChar === null || this.charAdded) {
      return;
    }
    this.word.splice(addBefore, 0, {
      id: Math.max(...this.word.map((el) => el.id)) + 1,
      char: this.selectedChar,
      initialChar: "",
      locked: false,
    });
    this.selectedChar = null;
    this.charAdded = true;
  }

  dragChar(event: DragEvent, id: string) {
    event.dataTransfer.setData("id", id);
  }

  trashElement(event: Event) {
    let id: number;
    if (event instanceof DragEvent) {
      id = +event.dataTransfer.getData("id");
    } else if (!isNaN(+this.selectedChar)) {
      id = +this.selectedChar;
    } else {
      return;
    }
    if (this.word.find((el) => el.id === id).locked) {
      return;
    }
    this.word = this.word.filter((el) => el.id !== id);
    this.charAdded = false;
  }
}
</script>

<style scoped>
.word-item {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}
.placeholder {
  min-width: 1em;
  border: 1px dashed black;
}
</style>
