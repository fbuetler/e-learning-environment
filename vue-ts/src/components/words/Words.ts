import words from "@/assets/words/words.json";

export type wordElement = {
  id: number;
  char: string;
  initialChar: string;
  locked: boolean;
};

export function LoadWords(key: string): [wordElement[], string[]] {
  const allWords = words;
  const keys = Object.keys(allWords[key]);
  const chosenWord = keys[Math.floor(Math.random() * keys.length)];
  const wordParts = chosenWord.split("");

  const similarWords = allWords[key][chosenWord];
  const word = new Array<wordElement>();
  for (let i = 0; i < wordParts.length; i++) {
    word.push({
      id: i,
      char: wordParts[i],
      initialChar: wordParts[i],
      locked: false,
    });
  }
  return [word, similarWords];
}
