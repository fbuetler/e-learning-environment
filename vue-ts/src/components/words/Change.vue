<template>
  <div @dragend.prevent="selectedChar = null">
    <div>
      Versuch ein neues Wort zu bilden, indem du einen Buchstaben Ã¤nderst.
    </div>
    <div class="word-container flex-item flex-row flex-center">
      <div v-for="element in word" :key="element.id">
        <div
          class="word-char card"
          :class="{ locked: element.locked }"
          @click="changeChar(element.id)"
          @dragover.prevent
          @dragend.prevent
          @drop.stop.prevent="changeChar(element.id)"
        >
          {{ element.char }}
        </div>
      </div>
    </div>
    <hr />
    <div
      class="interaction-container flex-item flex-row flex-center flex-stretch"
    >
      <Alphabet
        :selectedChar="selectedChar"
        @char-selected="selectedChar = $event"
      />
      <Undo @undo-operation="undo()" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "@/components/GameMixins";
import Alphabet from "@/components/words/Alphabet.vue";
import Undo from "@/components/Undo.vue";
import { LoadWords, wordElement } from "@/components/words/Words";

@Component<Change>({
  components: {
    Alphabet,
    Undo,
  },
})
export default class Change extends Mixins(GameMixin) implements GameInterface {
  word: wordElement[] = null;
  similarWords: string[] = null;
  selectedChar: string = null;

  isStarted(): boolean {
    return this.word === null;
  }

  restartGame() {
    [this.word, this.similarWords] = LoadWords("change", 1);
  }

  isCorrect(): boolean {
    return this.similarWords.includes(this.word.map((el) => el.char).join(""));
  }

  changeChar(id: number) {
    if (this.selectedChar === null) {
      return;
    }
    const element = this.word.find((el) => el.id === id);
    if (element.locked) {
      return;
    }
    if (this.selectedChar === element.initialChar) {
      element.char = element.initialChar;
      this.word.map((el) => (el.locked = false));
      return;
    }
    this.word.filter((el) => el.id !== id).forEach((el) => (el.locked = true));
    element.char = this.selectedChar;
  }

  undo() {
    this.word.forEach((el) => {
      el.char = el.initialChar;
      el.locked = false;
    });
    this.selectedChar = null;
  }
}
</script>
