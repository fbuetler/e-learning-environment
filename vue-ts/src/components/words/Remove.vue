<template>
  <div>
    <div class="word-container">
      <div v-for="element in word" :key="element.id">
        <div
          class="word-char"
          :class="{
            locked: element.locked,
            selected: element.id === selectedChar,
          }"
          v-if="element.char !== ''"
          @click="selectedChar = element.id"
          draggable
          @dragover.prevent
          @dragend.prevent
          @dragstart="selectedChar = element.id"
        >
          {{ element.char }}
        </div>
      </div>
    </div>
    <hr />
    <div class="interaction-container">
      <Trashcan @trashed-element="trashElement()" @click="trashElement()" />
      <Undo @undo-operation="undo($event)" />
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from "vue-property-decorator";
import GameMixin, { GameInterface } from "../Game";
import Undo from "@/components/Undo.vue";
import Trashcan from "@/components/Trashcan.vue";
import { LoadWords, wordElement } from "./Words";

@Component<Remove>({
  components: {
    Trashcan,
    Undo,
  },
})
export default class Remove extends Mixins(GameMixin) implements GameInterface {
  private dataKey = "remove";
  private word: wordElement[] = null;
  private similarWords: string[] = null;
  private selectedChar: number = null;
  private charRemoved = false;

  isStarted(): boolean {
    return this.word === null;
  }

  restartGame() {
    [this.word, this.similarWords] = LoadWords(this.dataKey);
    this.selectedChar = null;
    this.charRemoved = false;
  }

  isCorrect(): boolean {
    return this.similarWords.includes(this.word.map((el) => el.char).join(""));
  }

  trashElement() {
    if (this.charRemoved) {
      return;
    }
    this.word.forEach((el) => {
      if (el.id === this.selectedChar) {
        el.char = "";
      } else {
        el.locked = true;
      }
    });
    this.charRemoved = true;
  }

  undo() {
    this.word.forEach((el) => {
      el.char = el.initialChar;
      el.locked = false;
    });
    this.selectedChar = null;
    this.charRemoved = false;
  }
}
</script>
