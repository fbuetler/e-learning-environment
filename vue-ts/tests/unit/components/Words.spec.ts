import { shallowMount } from "@vue/test-utils";
import Add from "@/components/words/Add.vue";
import Change from "@/components/words/Change.vue";
import Remove from "@/components/words/Remove.vue";
import Swap from "@/components/words/Swap.vue";
import { wordElement } from "@/components/words/Words";

function prepareWord(word: string): wordElement[] {
  return word.split("").map((el, i) => {
    return { id: i, char: el, initialChar: el, locked: true };
  });
}

function addToWord(
  word: wordElement[],
  char: string,
  pos: number
): wordElement[] {
  word.splice(pos, 0, {
    id: Math.max(...word.map((el) => el.id)) + 1,
    char: char,
    initialChar: "",
    locked: false,
  });
  return word;
}

function copy(el) {
  return JSON.parse(JSON.stringify(el));
}

const rawWord = "ADLER";
const mockWord = prepareWord(rawWord);
const mockSimilarWords = ["RADLER", "TADLER", "ADLERN", "ADLERS"];
jest.mock("@/components/words/Words", () => ({
  LoadWords() {
    return [
      JSON.parse(JSON.stringify(mockWord)),
      JSON.stringify(mockSimilarWords).slice(),
    ];
  },
  findCorrectAndWrongSolutions(
    word: string,
    similarWord: string
  ): [number, string, number, string] {
    return [0, "", 0, ""];
  },
  get alphabet() {
    return ["A", "B", "C"];
  },
  get items() {
    return this.alphabet.map((el, i) => {
      return {
        id: i,
        type: i,
        value: i,
        img: el,
        class: "word-char",
      };
    });
  },
}));

describe("Add.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Add, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("mock used correctly", () => {
    expect(wrapper.vm["word"].map((el) => el.char).join("")).toBe(rawWord);
  });

  it("initially all chars are locked and nothing is added", () => {
    expect(wrapper.vm["word"].every((el) => el.locked)).toBeTruthy();
    expect(wrapper.vm["charAdded"]).toBeFalsy();
  });

  for (let i = 0; i <= mockWord.length; i++) {
    it(`adding a char by click at position ${i}`, async () => {
      await wrapper.setData({ selected: 1 }); // 'B'
      wrapper
        .findAll("rect")
        .at(i)
        .trigger("click");
      const newWord = addToWord(copy(mockWord), "B", i);
      expect(wrapper.vm["word"]).toEqual(newWord);
    });
  }

  it("no char selected", () => {
    wrapper.vm.addChar(0);
    expect(copy(wrapper.vm["word"])).toEqual(copy(mockWord));
  });

  it("add a char twice", async () => {
    await wrapper.setData({ selected: 1 }); // 'B'
    wrapper.vm.addChar(0);
    const word = copy(wrapper.vm["word"]);
    await wrapper.setData({ selected: 1 }); // 'B'
    wrapper.vm.addChar(0);
    expect(copy(wrapper.vm["word"])).toEqual(word);
  });

  it("correct word is evaluated as correct", async () => {
    await wrapper.setData({ word: prepareWord(mockSimilarWords[0]) });
    expect(wrapper.vm.isCorrect()).toBeTruthy();
  });

  it("incorrect word is evaluated as incorrect", async () => {
    await wrapper.setData({ word: prepareWord("TESTTEST") });
    expect(wrapper.vm.isCorrect()).toBeFalsy();
  });

  for (let i = 0; i <= mockWord.length; i++) {
    it(`undo is handled correctly at position ${i}`, async () => {
      await wrapper.setData({ word: addToWord(copy(mockWord), "B", i) });
      wrapper.vm.undo();
      expect(copy(wrapper.vm["word"])).toEqual(mockWord);
      expect(wrapper.vm["charAdded"]).toBeFalsy();
    });
  }
});

describe("Change.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Change, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  // TODO: add more tests
});

describe("Remove.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Remove, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  // TODO: add more tests
});

describe("Swap.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Swap, {});
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("renders correctly", () => {
    // TODO: problems with mocking multiple functions in Words
  });

  // TODO: add more tests
});
