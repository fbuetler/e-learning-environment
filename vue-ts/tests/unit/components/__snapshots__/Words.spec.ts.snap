// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add.vue renders correctly 1`] = `
<div>
   
  <div>
    
    Versuch ein neues Wort zu bilden, indem du einen Buchstaben hinzufügst.
  
  </div>
   
  <div
    class="word-container flex-item flex-col flex-center flex-flex"
    id="word-container"
  >
    <div
      class="flex-item flex-row flex-center"
    >
      <div
        class="word-char card locked"
        id="word-char-0"
      >
        
        A
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-1"
      >
        
        D
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-2"
      >
        
        L
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-3"
      >
        
        E
      
      </div>
      <div
        class="word-char card locked"
        id="word-char-4"
      >
        
        R
      
      </div>
       
      <div />
    </div>
     
    <div
      class="svg-container"
    >
      <svg
        class="svg-item"
      >
        <defs>
          <marker
            id="arrowhead"
            markerHeight="6"
            markerWidth="6"
            orient="auto"
            refX="1"
            refY="5"
            viewBox="0 0 10 10"
          >
            <path
              d="M 0 5 L 10 0 L 10 10 z"
            />
          </marker>
        </defs>
         
        <line
          fill="transparent"
          id="arrow-0"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-1"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-2"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-3"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-4"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
        <line
          fill="transparent"
          id="arrow-5"
          marker-start="url(#arrowhead)"
          stroke="black"
          stroke-width="3"
          x1="0"
          x2="0"
          y1="10"
          y2="50"
        />
         
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-0"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-1"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-2"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-3"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-4"
          width="40"
          x="-20"
          y="-10"
        />
        <rect
          fill="transparent"
          height="0"
          id="rect-around-arrow-5"
          width="40"
          x="-20"
          y="-10"
        />
      </svg>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch"
  >
    <itemselection-stub
      items="[object Object],[object Object],[object Object]"
    />
     
    <undo-stub />
  </div>
</div>
`;

exports[`Change.vue renders correctly 1`] = `
<div>
  <div>
    
    Versuch ein neues Wort zu bilden, indem du einen Buchstaben änderst.
  
  </div>
   
  <div
    class="word-container flex-item flex-row flex-center"
  >
    <div>
      <div
        class="word-char card locked"
      >
        
        A
      
      </div>
    </div>
    <div>
      <div
        class="word-char card locked"
      >
        
        D
      
      </div>
    </div>
    <div>
      <div
        class="word-char card locked"
      >
        
        L
      
      </div>
    </div>
    <div>
      <div
        class="word-char card locked"
      >
        
        E
      
      </div>
    </div>
    <div>
      <div
        class="word-char card locked"
      >
        
        R
      
      </div>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch"
  >
    <itemselection-stub
      items="[object Object],[object Object],[object Object]"
    />
     
    <undo-stub />
  </div>
</div>
`;

exports[`Remove.vue renders correctly 1`] = `
<div>
  <div>
    
    Versuch ein neues Wort zu bilden, indem du einen Buchstaben entfernst.
  
  </div>
   
  <div
    class="word-container flex-item flex-row flex-center"
  >
    <div>
      <div
        class="word-char card clickable locked"
        draggable="true"
      >
        
        A
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable locked"
        draggable="true"
      >
        
        D
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable locked"
        draggable="true"
      >
        
        L
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable locked"
        draggable="true"
      >
        
        E
      
      </div>
    </div>
    <div>
      <div
        class="word-char card clickable locked"
        draggable="true"
      >
        
        R
      
      </div>
    </div>
  </div>
   
  <hr />
   
  <div
    class="interaction-container flex-item flex-row flex-center flex-stretch"
  >
    <trashcan-stub />
     
    <undo-stub />
  </div>
</div>
`;
